### PR TITLE
feat(browserstack-service): add skipAppOverride capability (v8 backport)

### PR DESCRIPTION
**Pull request by mandeep1108**
## Proposed changes

Adds a `skipAppOverride` service option to `@wdio/browserstack-service` for App Automate (Appium) runs.

Today the service always manages the app: it uploads the file given in the `app` option and injects the
resulting `appium:app` capability. Some teams manage the app themselves (a pre-uploaded `bs://` hash supplied
directly as a driver capability, or via `BROWSERSTACK_APP_ID`) and don't want the service to upload or override it.

With `skipAppOverride: true`, the service:
1. Classifies the session as **App Automate** even when no `app` option is set.
2. Does **not** upload an app.
3. Does **not** inject an `appium:app` capability — the user supplies the app reference themselves.
4. Emits a one-time warning so the behaviour is explicit.

Edge cases:
- `skipAppOverride: true` **with** an `app` option → a conflict warning is logged and the `app` option is ignored
  (not uploaded/injected); the run proceeds as App Automate.
- `skipAppOverride: false` **with no** `app` → the run fails fast with a clear configuration error before any
  session starts (prevents a silent misconfiguration).
- Unset → existing behaviour is unchanged.

The option is never forwarded to the hub in `bstack:options` (added to the internal cap strip-list).

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Polish / Bugfix / Breaking change / Documentation update / Specification changes / Internal updates

## Checklist

- [x] I have read the CONTRIBUTING doc
- [x] I have added tests that prove my feature works (`tests/skipAppOverride.test.ts`; updated `tests/cli/modules/automateModule.test.ts` mock for the new `isTrue` usage)
- [x] I have added proper type definitions (`skipAppOverride?: boolean` on `BrowserstackConfig`, with JSDoc)
- [ ] I have added the necessary documentation (if appropriate)

## Backport Request

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [x] Backport of the v9 change in webdriverio/webdriverio#15373 (this PR targets the `v8` branch).

## Further comments

Files changed (all in `packages/wdio-browserstack-service`):
- `src/types.ts` — declare the `skipAppOverride?: boolean` option.
- `src/config.ts` — App Automate classification honors the option.
- `src/service.ts` — the worker-local `_isAppAutomate()` honors the option (no injected app cap required).
- `src/util.ts` — new `validateSkipAppOverride()` (warning + the three edge cases).
- `src/launcher.ts` — invoke the validator once in `onPrepare`; the `false`+no-app case surfaces as `SevereServiceError`.
- `src/constants.ts` — add `skipAppOverride` to the capabilities strip-list.
- `src/cli/modules/automateModule.ts` — session name/status endpoint routing honors the option.

Verification:
- **Unit:** `tsc --noEmit` clean; vitest green — `skipAppOverride.test.ts`, `config.test.ts`, `util.test.ts`, `service.test.ts`.
- **On-device (App Automate, Samsung Galaxy S23 / Android 13):**
  - `skipAppOverride: true` with the app supplied via an `appium:app` cap → App Automate session created, one-time warning, **no upload**.
  - `skipAppOverride: true` + an `app` option → conflict warning, the `app` option ignored in favour of the user's cap, no upload, session created.
  - `skipAppOverride: false` with no app → fails fast with the config error before any session starts.

### CODE REVIEW

File: packages/wdio-browserstack-service/src/launcher.ts [hunk 1]
```diff
@@ -219,6 +220,15 @@ export default class BrowserstackLauncherService implements Services.ServiceInst
...
+        // else. Runs once here in the launcher (main process). Edge-2 (explicit false + no app) is a
+        // deliberate pre-session config error, surfaced as SevereServiceError so the run aborts cleanly.
+        try {
+            validateSkipAppOverride(this._options)
+        } catch (error) {
+            throw new SevereServiceError((error as Error).message)
+        }
+
         // // Send Funnel start request
         await sendStart(this.browserStackConfig)
```
[the remark this answers is not in the document]
Comment (mandeep1108) [reply packages/wdio-browserstack-service/src/launcher.ts:236]: Fixed — after the `validateSkipAppOverride` call in `onPrepare` we now reset `this.browserStackConfig.app = this._options.app`, so the singleton stays consistent when edge-1 clears the app. Applied to both the v8 branch (4c901e6c0) and the v9 PR #15373 (3e67ad7f8).

File: packages/wdio-browserstack-service/src/util.ts [hunk 2]
```diff
@@ -1389,6 +1389,50 @@ export function isFalse(value?: any) {
...
+        return
+    }
+
+    // Edge 2: app not specified + skipAppOverride explicitly false → pre-session config error.
+    if (isFalse(skipAppOverride) && isUndefined((options as BrowserstackConfig).app)) {
+        throw new Error('App capability is missing. When skipAppOverride is set to \'false\', a valid app capability (hash/shareable id/path/custom_id) must be provided.')
+    }
+
+    // Edge 3 (and default): unset + no app → unchanged behavior (no-op).
+}
```
[the remark this answers is not in the document]
Comment (mandeep1108) [reply packages/wdio-browserstack-service/src/util.ts:1437]: Leaving as-is: `validateSkipAppOverride` is invoked exactly once per run from the launcher's `onPrepare` (main process), so the repeated-call path where the conflict warning would arrive without the standard preamble can't occur here. The app-clear is idempotent regardless. Happy to revisit if the helper is ever reused outside onPrepare.

**Comment by mandeep1108** [comment]
Reviewed this backport — logic is consistent across all three call sites (launcher, per-worker service, CLI session-marking) and mirrors the v9 PR (#15373). Test coverage looks thorough: all three edge cases, string coercion, emit-once, cloud-leak strip, and endpoint-routing assertions. ✅ LGTM.

One non-blocking nit — `src/util.ts` docstring on `validateSkipAppOverride`:
- The conflict-warning string is intentionally *not* verbatim vs the other SDKs (WDIO reads `app` from `wdio.conf.js` service options, not `browserstack.yml`), but the comment claims verbatim parity — worth flagging the exception so a future maintainer doesn't "fix" it back.
- The cited line numbers drift with unrelated edits; referencing the method name instead would be more durable.

**Review by mandeep1108 — COMMENTED** [review]
Reviewed this backport — logic is consistent across the launcher, per-worker service, and CLI session-marking paths, and mirrors the v9 PR (#15373). Test coverage looks thorough: all three edge cases, string coercion, emit-once, cloud-leak strip, and endpoint-routing assertions. ✅ LGTM. One non-blocking docstring nit inline.

File: packages/wdio-browserstack-service/src/util.ts (hunk 2 above)
Comment (mandeep1108) [inline packages/wdio-browserstack-service/src/util.ts:1437]: Non-blocking nit: the conflict-warning string here reads `app` from the `wdio.conf.js` service options rather than a `browserstack.yml` path, so its wording is intentionally WDIO-specific. Might be worth a note in the docstring so a future maintainer doesn't assume it should match the other-SDK phrasing verbatim. Also, hard-coded line-number references tend to drift with unrelated edits — a method/function name would be more durable.

**Comment by mandeep1108** [comment]
Thanks! Addressed the docstring nit in `7342afc1b`: the `validateSkipAppOverride` docstring now explicitly flags that the edge-1 conflict wording is *intentionally* WDIO-specific ("browserstack service options", not "browserstack.yml") so a future maintainer won't "fix" it back, and drops the drifting line-number references in favor of the method name. Also backported the v9 review fixes (InsightsHandler skipAppOverride + app-supply doc correction + tests) here to keep the two PRs in lockstep.

File: packages/wdio-browserstack-service/src/util.ts (hunk 2 above)
Comment (mandeep1108) [reply packages/wdio-browserstack-service/src/util.ts:1437]: Done in `7342afc1b` — the `validateSkipAppOverride` docstring now explicitly notes the edge-1 conflict wording is intentionally WDIO-specific ("browserstack service options", not "browserstack.yml") so it isn't "fixed" back to match the other SDKs, and the drifting line-number references were replaced with the method name (`BrowserStackJavaAgent#processAppOptions`).

**Comment by dprevost-LMI** [comment]
`wdio-browserstack-service` moved to https://github.com/browserstack/wdio-browserstack-service; I suggest moving the PR there too.
